### PR TITLE
[Feature] Add FlashInfer cuDNN backend for ViT attention

### PR DESCRIPTION
**Pull request by dorhuri123**
## Summary
- Adds cuDNN-accelerated attention via FlashInfer's `cudnn_batch_prefill_with_kv_cache` for Vision Transformer encoders
- Supports bf16 with attention masks, targeting Qwen2.5-VL and Qwen3-VL models
- Full integration for Qwen3-VL; parameter plumbing for Qwen2.5-VL (full integration in follow-up)

### New cuDNN wrapper (`vllm/v1/attention/ops/vit_attn_wrappers.py`)
- `flashinfer_wrapper()` calls FlashInfer's cuDNN batch prefill API
- Uses `actual_seq_lens_q/kv` for variable-length sequences (same approach as MLA attention)
- Handles 3-section `cu_seqlens` format (batch_offsets_qk, batch_offsets_v, batch_offsets_o)
- Registered as custom op for `torch.compile` compatibility

### MMEncoderAttention updates (`vllm/model_executor/layers/attention/mm_encoder_attention.py`)
- New `_forward_flashinfer()` method with FlashInfer dispatch
- `workspace_buffer` parameter for pre-allocated cuDNN workspace (128MB)
- `sequence_lengths` parameter threaded through all forward methods

### Qwen3-VL integration (`vllm/model_executor/models/qwen3_vl.py`)
- Batch bucket padding (`BATCH_BUCKETS = [8, 16, 32, 64]`) for cuDNN graph caching
- `compute_flashinfer_cu_seqlens()` computes the 3-section cu_seqlens format
- Fixed `max_seqlen` to 128K to avoid cuDNN recompilation
- FlashInfer added to supported backends

### Platform support (`vllm/platforms/cuda.py`)
- `FLASHINFER` added to CUDA's `get_supported_vit_attn_backends()`

## Usage
```bash
# Via CLI server
vllm serve Qwen/Qwen3-VL-4B-Instruct --mm-encoder-attn-backend FLASHINFER

# Via Python API
llm = LLM(model="Qwen/Qwen3-VL-4B-Instruct", mm_encoder_attn_backend="FLASHINFER")
```

## Credits
Based on [CentML/vllm#30](https://github.com/CentML/vllm/pull/30).

## Benchmark Results — H100 SXM 80GB

**Setup:** Qwen3-VL-4B-Instruct, bf16, max_model_len=4096, FlashInfer 0.6.3, CUDA 12.8

### Correctness
Both backends produce equivalent descriptions of the same test image (translucent dice):

| Backend | Output |
|---------|--------|
| FLASH_ATTN | *"This is a 3D illustration of four colorful, translucent dice—red, blue, green, and yellow—floating against a plain white background. The red die is in the foreground and slightly blurred, while the others are in the background, creating a sense of depth."* |
| FLASHINFER | *"This is a 3D illustration of four colorful, translucent dice—red, blue, green, and yellow—floating in mid-air against a plain white background. The dice are slightly blurred, suggesting motion or depth, and are rendered with a glossy, glass-like finish."* |

**Result: PASS** — Both backends produce semantically identical, correct output. Minor wording differences are expected due to floating-point non-determinism between attention implementations.

### Performance — Text (50 prompts, random dataset, `vllm bench serve`)

| Metric | FLASH_ATTN | FLASHINFER | Delta |
|--------|-----------|------------|-------|
| Request throughput (req/s) | 5.34 | 5.27 | -1.3% |
| Output token throughput (tok/s) | 684.16 | 674.26 | -1.4% |
| Total token throughput (tok/s) | 6,157.40 | 6,068.38 | -1.4% |
| Mean TTFT (ms) | 4,250.93 | 4,348.84 | +2.3% |
| Mean TPOT (ms) | 5.33 | 5.35 | +0.4% |
| Mean ITL (ms) | 5.33 | 5.35 | +0.4% |
| P99 ITL (ms) | 14.36 | 6.75 | **-52.9%** |

**Result:** No regression. Performance is within noise margin (~1-2%). Notably, P99 inter-token latency improved by 53% with FlashInfer. The benchmark uses random text prompts; ViT-heavy workloads (many images) would show larger differences in the encoder attention path.

### Performance — Images (50 prompts, random-mm dataset, 1 image/request, `vllm bench serve`)

| Metric | FLASH_ATTN | FLASHINFER | Speedup |
|--------|-----------|------------|---------|
| Request throughput (req/s) | 0.71 | 5.22 | **7.4x** |
| Output token throughput (tok/s) | 91.38 | 667.65 | **7.3x** |
| Total token throughput (tok/s) | 822.43 | 6,008.85 | **7.3x** |
| Mean TTFT (ms) | 28,268 | 6,673 | **4.2x faster** |
| Mean TPOT (ms) | 33.79 | 22.46 | **1.5x faster** |
| Benchmark duration (s) | 70.04 | 9.59 | **7.3x faster** |

**Result: 7.3x throughput improvement on image workloads with cuDNN ViT attention.**

### 1. Verify FlashInfer cuDNN is available
```bash
python -c "
import torch
print('GPU:', torch.cuda.get_device_name(0))
print('CUDA:', torch.version.cuda)
from flashinfer.prefill import cudnn_batch_prefill_with_kv_cache
print('FlashInfer cuDNN: OK')
from vllm.platforms import current_platform
print('ViT backends:', current_platform.get_supported_vit_attn_backends())
"
```

### 2. Correctness test — FLASH_ATTN baseline
```bash
python -c "
from vllm import LLM, SamplingParams

llm = LLM(
    model='Qwen/Qwen3-VL-4B-Instruct',
    dtype='bfloat16',
    max_model_len=4096,
    max_num_seqs=4,
    limit_mm_per_prompt={'image': 1},
    mm_encoder_attn_backend='FLASH_ATTN',
    allowed_local_media_path='/tmp',
)

messages = [{'role': 'user', 'content': [
    {'type': 'image_url', 'image_url': {'url': 'file:///tmp/test_image.png'}},
    {'type': 'text', 'text': 'Describe this image briefly.'},
]}]

outputs = llm.chat(messages=[messages], sampling_params=SamplingParams(temperature=0, max_tokens=64))
print('FLASH_ATTN output:', outputs[0].outputs[0].text)
"
```

### 3. Correctness test — FLASHINFER cuDNN
```bash
python -c "
from vllm import LLM, SamplingParams

llm = LLM(
    model='Qwen/Qwen3-VL-4B-Instruct',
    dtype='bfloat16',
    max_model_len=4096,
    max_num_seqs=4,
    limit_mm_per_prompt={'image': 1},
    mm_encoder_attn_backend='FLASHINFER',
    allowed_local_media_path='/tmp',
)

messages = [{'role': 'user', 'content': [
    {'type': 'image_url', 'image_url': {'url': 'file:///tmp/test_image.png'}},
    {'type': 'text', 'text': 'Describe this image briefly.'},
]}]

outputs = llm.chat(messages=[messages], sampling_params=SamplingParams(temperature=0, max_tokens=64))
print('FLASHINFER output:', outputs[0].outputs[0].text)
"
```

### 4. Serving benchmark — FLASH_ATTN
```bash
python -m vllm.entrypoints.openai.api_server \
    --model Qwen/Qwen3-VL-4B-Instruct \
    --dtype bfloat16 \
    --max-model-len 4096 \
    --max-num-seqs 4 \
    --limit-mm-per-prompt '{"image": 1}' \
    --mm-encoder-attn-backend FLASH_ATTN \
    --port 8000 &

sleep 90

vllm bench serve \
    --model Qwen/Qwen3-VL-4B-Instruct \
    --backend vllm \
    --port 8000 \
    --num-prompts 50 \
    --endpoint /v1/completions

kill %1
```

### 5. Serving benchmark — FLASHINFER
```bash
python -m vllm.entrypoints.openai.api_server \
    --model Qwen/Qwen3-VL-4B-Instruct \
    --dtype bfloat16 \
    --max-model-len 4096 \
    --max-num-seqs 4 \
    --limit-mm-per-prompt '{"image": 1}' \
    --mm-encoder-attn-backend FLASHINFER \
    --port 8000 &

sleep 90

vllm bench serve \
    --model Qwen/Qwen3-VL-4B-Instruct \
    --backend vllm \
    --port 8000 \
    --num-prompts 50 \
    --endpoint /v1/completions

kill %1
```

### 6. Serving image benchmark — FLASH_ATTN
```bash
vllm serve Qwen/Qwen3-VL-4B-Instruct \
    --mm-encoder-attn-backend FLASH_ATTN \
    --max-model-len 4096 \
    --gpu-memory-utilization 0.9 \
    --port 8000 &

sleep 120

vllm bench serve \
    --backend openai-chat \
    --model Qwen/Qwen3-VL-4B-Instruct \
    --dataset-name random-mm \
    --random-mm-base-items-per-request 1 \
    --num-prompts 50 \
    --endpoint /v1/chat/completions \
    --port 8000

kill %1
```

### 7. Serving image benchmark — FLASHINFER
```bash
vllm serve Qwen/Qwen3-VL-4B-Instruct \
    --mm-encoder-attn-backend FLASHINFER \
    --max-model-len 4096 \
    --gpu-memory-utilization 0.9 \
    --port 8000 &

sleep 120

vllm bench serve \
    --backend openai-chat \
    --model Qwen/Qwen3-VL-4B-Instruct \
    --dataset-name random-mm \
    --random-mm-base-items-per-request 1 \
    --num-prompts 50 \
    --endpoint /v1/chat/completions \
    --port 8000

kill %1
```

### Results Summary
- [x] Verify Qwen3-VL inference with `--mm-encoder-attn-backend FLASHINFER` — **PASS**
- [x] Check no regression with default backends (FLASH_ATTN) — **PASS, no regression**
- [x] Correctness comparison: FlashInfer vs Flash Attention produce equivalent outputs — **PASS**
- [x] Serving benchmark on H100: no throughput/latency regression — **PASS**
- [x] Serving image benchmark on H100: **7.3x throughput improvement** — **PASS**

### CODE REVIEW

File: vllm/v1/attention/ops/vit_attn_wrappers.py [hunk 1]
```diff
@@ -191,3 +191,88 @@ def vit_torch_sdpa_wrapper(
...
+        reshape_batch_size = q.shape[0]
+        q, k, v = (einops.rearrange(x, "b s ... -> (b s) ...") for x in [q, k, v])
+
+    assert len(cu_seqlens) % 3 == 0, "cu_seqlens must be divisible by 3"
+    cu_seqlength = len(cu_seqlens) // 3
+    batch_offsets_qk = cu_seqlens[:cu_seqlength].view(-1, 1, 1, 1)
+    batch_offsets_v = cu_seqlens[cu_seqlength : cu_seqlength * 2].view(-1, 1, 1, 1)
+    batch_offsets_o = cu_seqlens[cu_seqlength * 2 :].view(-1, 1, 1, 1)
+    sequence_lengths = sequence_lengths.view(-1, 1, 1, 1)
+    max_seqlen = max_seqlen.item()
```
[the remark this answers is not in the document]
Comment (dorhuri123) [reply vllm/v1/attention/ops/vit_attn_wrappers.py]: This is safe — flashinfer_wrapper is registered as a custom op via direct_register_custom_op. torch.compile uses the fake impl and never traces into this function.

File: vllm/v1/attention/ops/vit_attn_wrappers.py [hunk 2]
```diff
@@ -191,3 +191,88 @@ def vit_torch_sdpa_wrapper(
...
+        q,
+        k,
+        v,
+        scale,
+        workspace_buffer,
+        max_token_per_sequence=max_seqlen,
+        max_sequence_kv=max_seqlen,
+        actual_seq_lens_q=sequence_lengths,
+        actual_seq_lens_kv=sequence_lengths,
+        causal=False,
```
[the remark this answers is not in the document]
Comment (dorhuri123) [reply vllm/v1/attention/ops/vit_attn_wrappers.py:331]: Confirmed — ViT attention is always non-causal. Every patch attends to all other patches bidirectionally.

File: vllm/model_executor/layers/attention/mm_encoder_attention.py [hunk 3]
```diff
@@ -199,7 +230,8 @@ def forward_cpu(
         key: torch.Tensor,
         value: torch.Tensor,
         cu_seqlens: torch.Tensor | None = None,
-        max_seqlen: torch.Tensor | None = None,  # Only used for Flash Attention
+        max_seqlen: torch.Tensor | None = None,
+        sequence_lengths: torch.Tensor | None = None,
```
[the remark this answers is not in the document]
Comment (dorhuri123) [reply vllm/model_executor/layers/attention/mm_encoder_attention.py:272]: Intentional — the caller passes sequence_lengths via **kwargs through CustomOp.forward() (line 125-126 in custom_op.py), which blindly forwards all arguments. If forward_cpu doesn't accept it, it would crash with TypeError at runtime. Same established pattern as max_seqlen which is also accepted here but only used by Flash Attention.

File: vllm/model_executor/layers/attention/mm_encoder_attention.py [hunk 4]
```diff
@@ -209,7 +241,8 @@ def forward_xpu(
...
```
[the remark this answers is not in the document]
Comment (dorhuri123) [reply vllm/model_executor/layers/attention/mm_encoder_attention.py:283]: Same as above — required to avoid TypeError when CustomOp dispatches to this method.
